### ISSUE TITLE
Collect observability data on AWS Lambda

### ISSUE THREAD

**Issue by binarylogic** [timberio/vector#874]
An idea I've been thinking about us deploying Vector as an [AWS Lambda Layer](https://docs.aws.amazon.com/lambda/latest/dg/configuration-layers.html), which would collect logs, metrics, etc, and ship them right from the Lambda function. This would server users that do not want to use AWS CloudWatch to collect this data.

**Comment by a-rodin** [comment]
When Vector is used with certain sinks like S3, it takes long time to shut it down (around 40 seconds). Might it be a problem with short-living invocations?

**Comment by loony-bean** [comment]
I'm not sure if the official runtime does the same, but in [this (already archived) project](https://github.com/srijs/rust-aws-lambda) the process is described as long living

> This project forgoes embedding and instead leverages lambda's official Go support. With Go, lambda runs a standard Linux binary containing a server and then sends it gob-encoded messages via rpc. This project reimplements all that machinery in Rust, using rust-gob for gob support and a custom tokio server runtime. Lambda does not care that the Linux binary it runs is written in Rust rather than Go as long as they behave the same.

**Comment by binarylogic** [comment]
> When Vector is used with certain sinks like S3, it takes long time to shut it down (around 40 seconds). Might it be a problem with short-living invocations?

That's odd. A broader question is why does Vector take that long to shut down? I assume when a `SIGTERM` is received we'd flush immediately and start the shutdown process. I expect for that to take _some_ time, but not 40 seconds...

Is that worth opening another issue to investigate?

**Comment by lukesteensen** [comment]
Yeah, we should at least try to flush immediately when we get a shutdown signal. Unless S3 itself is being super slow, there's no reason it should be taking that long to finish.

**Comment by LucioFranco** [comment]
I think sink flushing is generally pretty quick, we'd have to see a flamegraph or something to determine what really is the issue. I have seen some of our "blocking" based sources have trouble shutting down.

**Comment by loony-bean** [comment]
FWIW I've observed this behaviour only when some services are not healthy

```sh
Sep 18 10:12:48.117 ERROR healthcheck{name=cloudwatch}: vector::topology::builder: Healthcheck: Failed Reason: deadline has elapsed
Sep 18 10:13:08.137 ERROR sink{name="cloudwatch"}: vector::sinks::util: request failed. error=Couldn't find AWS credentials in environment, credentials file, or IAM role.
Sep 18 10:13:25.946  INFO vector::topology: Shutting down... Waiting on: cloudwatch. 54 seconds left
Sep 18 10:13:30.945  INFO vector::topology: Shutting down... Waiting on: cloudwatch. 49 seconds left
```

**Comment by wellbarreto** [comment]
Up
